Issue 30 find partial page citations

mattdahl opened the pull request:
This enables `eyecite` to find citations that don't yet have a final page number, but just have a placeholder `___` space instead. This is common with the `U.S.` reporter. This simple parsing change is the first half of addressing #30; figuring out how to match these partial citations will have to be done in CL.

Two things I thought of but didn't include:
1) Should we add a metadata flag to the citation objects like `has_proper_page_number` or something? This would be `False` if the page number is all underscores, `True` otherwise. I wasn't sure if this would actually be helpful or not.
2) Are there other characters besides underscores that courts use to indicate a placeholder for an eventual page number?

mlissner commented on tests/test_ResolveTest.py:
```diff
@@ -241,6 +241,14 @@ def test_id_resolution(self):
         )
+        # Test resolving an Id. citation with a pin cite when the previous
+        # citation only has a placeholder page. We expect this to still resolve
+        # because it's possible to pin cite to an opinion with pending final
```
Really? You can pin site using a page before you have a citation with the page? That's truly weird.

mattdahl replied:
Hmm, maybe not actually, I thought that people would just pin cite to whatever the document page number is (starting at 1) if the real page number was not available, but now I'm not sure. Is it better to expect this test to not resolve the Id. citation?

mattdahl replied:
Fixed in 6048422

mlissner commented on tests/test_FindTest.py:
```diff
@@ -181,8 +181,11 @@ def test_find_citations(self):
             ('1 U.S. f24601', []),
+            # Test with page number that is indicated as missing
+            ('1 U.S. ___',
+             [case_citation(volume='1', reporter='U.S.', page='___')]),
```
Do we use None for the `page` attribute in any other context? Maybe it's better to normalize these underscore things to `None` and then that answers your question about whether we need to add an extra attribute to the response to indicate this situation. What do you think?

mattdahl replied:
That's a good idea, I don't think we use `None` in the `page` attribute at all yet!

mattdahl replied:
Fixed in a120ca8

mlissner reviewed (changes requested):
Wow, this looks simple enough, thanks Matt! 

@jcushman do you have time to take a look and chime in before we merge? 

(A couple questions/comments in the review.)

jcushman commented:
> @jcushman do you have time to take a look and chime in before we merge?

Not with anything well informed. :)

For Id. cites I imagine they tend to also give a parallel cite with page numbers, e.g. S.Ct. or a docket number, and you could cite to that? Like "Foo v. Bar, 1 S.Ct. 3, 4 U.S. __ (2000). ... Id. at (something)". I'm not sure how that's handled, though. It wouldn't be a terrible idea to grep through for 2 or 3 underscores and see how these are actually used. Possibly one answer is there's always a parallel cite that's better for resolution of the citation.

mlissner commented:
I see you working, but let me know when this is ready for another review.

mattdahl commented:
> For Id. cites I imagine they tend to also give a parallel cite with page numbers, e.g. S.Ct. or a docket number, and you could cite to that? Like "Foo v. Bar, 1 S.Ct. 3, 4 U.S. __ (2000). ... Id. at (something)". I'm not sure how that's handled, though. It wouldn't be a terrible idea to grep through for 2 or 3 underscores and see how these are actually used. Possibly one answer is there's always a parallel cite that's better for resolution of the citation.

I found this to be sometimes the case but often the Id. cite is just `Id., at ___`, so I changed our behavior to just not match those. Enhancing this with the parallel cite info could be done as a part of #76 perhaps.

> Sounds good to me.

Hashability changes done in 08b9625.

----

N.B. I also had to make a change to the joke cite (5ee4145) because it was causing an issue with the `CitationBase.__post_init__` method for some reason after I implemented the missing page normalization logic.

mlissner reviewed (commented):
A couple little things, but looks close.

mattdahl commented on eyecite/models.py:
```diff
@@ -83,6 +83,12 @@ def __post_init__(self):
+            "page" in self.groups
+            and self.groups["page"] == len(self.groups["page"]) * "_"
+        ):
+            self.groups["page"] = None
```
Good point, I thought my way was faster, but I just tested it and the difference is negligible. Fixed in [d9966ce](https://github.com/freelawproject/eyecite/pull/116/commits/d9966cea0438f149b547c43cff19b4f5f27e2e15)

mlissner commented:
very nice, thank you!
